src/wp-includes/class-wp-query.php

```diff
@@ -3156,12 +3211,31 @@ public function get_posts() {
 			$where    = $clauses['where'] ?? '';
 			$groupby  = $clauses['groupby'] ?? '';
 			$join     = $clauses['join'] ?? '';
-			$orderby  = $clauses['orderby'] ?? '';
+			// Preserve orderby from posts_orderby_request if posts_clauses_request doesn't provide one.
+			$orderby  = $clauses['orderby'] ?? $orderby;
 			$distinct = $clauses['distinct'] ?? '';
 			$fields   = $clauses['fields'] ?? '';
 			$limits   = $clauses['limits'] ?? '';
 		}
 
+		/*
+		 * Ensure deterministic ordering to prevent duplicate records across pages.
+		 * Add ID tie-breaker after filters have been applied, so filters receive
+		 * the original orderby value (for backward compatibility) and the tie-breaker
+		 * is preserved even if filters modify the orderby.
+		 * 
+		 * Note: this is to circumvent a bug that is currently being tracked in
+		 * https://core.trac.wordpress.org/ticket/44349.
+		 */
+		if ( ! empty( $orderby ) && $deterministic_orderby_meta['needed'] ) {
+			// Check if ID tie-breaker is already present in the orderby string.
+			$id_tie_breaker_pattern = '/\b' . preg_quote( $wpdb->posts, '/' ) . '\.ID\b/i';
+			if ( ! preg_match( $id_tie_breaker_pattern, $orderby ) ) {
+				// Add ID as tie-breaker at the end.
+				$orderby .= ', ' . "{$wpdb->posts}.ID " . $deterministic_orderby_meta['order'];
+			}
+		}
```

Comment:
Based on the experience in https://core.trac.wordpress.org/ticket/64390 I've moved the logic beneath the filters